frontend/src/services/web3-provider/walletconnect-provider.ts

```diff
@@ -1,27 +1,46 @@
 import { hexValue } from 'ethers/lib/utils';
 
-import { EthereumProvider } from '@/services/web3-provider/ethereum-provider';
-import type { Eip1193Provider } from '@/services/web3-provider/types';
+import type { Eip1193Provider } from '@/services/web3-provider';
+import { EthereumProvider } from '@/services/web3-provider';
 import { WalletConnect } from '@/services/web3-provider/util-export';
 
+const BEAMER_PROJECT_ID = '8d13d0b06a0afde16d74f8214e4ba355';
+
 export async function createWalletConnectProvider(rpcList: {
   [chainId: string]: string;
 }): Promise<WalletConnectProvider | undefined> {
-  const provider = new WalletConnect({
-    rpc: rpcList,
-  }) as typeof WalletConnect;
-
-  await provider.enable();
-
-  if (provider.connected) {
-    const walletConnectProvider = new WalletConnectProvider(
-      provider as unknown as Eip1193Provider,
-    );
-    await walletConnectProvider.init();
-    return walletConnectProvider;
-  }
+  /* 
+    Due to a bug in @walletconnect/ethereum-provider (https://github.com/WalletConnect/walletconnect-monorepo/issues/1957)
+    we are wrapping the code in a promise with a self-invokable async function
+    so we can shortcut the lack of resolution in the `provider.enable()` promise
+  */
+
+  return new Promise((resolve) => {
+    (async () => {
+      const provider = await WalletConnect.init({
+        chains: Object.keys(rpcList).map((chainId) => parseInt(chainId)),
+        projectId: BEAMER_PROJECT_ID,
+        rpcMap: rpcList,
+      });
+
+      // Fixes the lack of promise resolution in `provider.enable()` when modal is closed
```

Comment:
I think it would be a bit cleaner if only the subscribeModal function would be wrapped in a promise (would remove the Promise on the top and the following anonymous async function). It could then be awaited in combination with provider.enable by using Promise.race. Could look like this:

```
    const modalClosed = new Promise((resolve) => {
      provider.modal?.subscribeModal(async (newState) => {
        if (!newState.open && !provider.session) {
          resolve(undefined);
        }
      });
    });
    const enabled = provider.enable();

    await Promise.race([enabled, modalClosed];
```

Only change this if you like. Your approach completely works. It looks just a bit confusing to me to have an anonymous async function wrapped in a Promise.

Reply:
Good idea and a clean one. Will switch to using `Promise.race`, thanks for the suggestion.